Combobox complex options

it-vegard opened the pull request:
### Description

Allow Combobox options as a list of objects, to support richer Options and separating the displayed label and the programmatical value used internally.

### Change summary

Add support for Combobox.options as `{ label: "Some label", value: "a value" }` in addition to the old `string[]` options.

it-vegard committed:
Remove unused function

it-vegard committed:
Support for complex options, using objects with label and value instead of strings

it-vegard committed:
Better typing for isTextInSelectedOptions value (boolean instead of truthy/falsy)

it-vegard committed:
Stricter value type: Only accept string values. This will work better with the allowNewValues prop, and simplify returning a value from onToggleSelected.

it-vegard committed:
More JSDoc

it-vegard committed:
Add test for complex options with label and value

KenAJoh committed:
Merge branch 'main' into combobox-complex-options

it-vegard committed:
Add example for complex options on the aksel-website

it-vegard committed:
yarn.lock

it-vegard committed:
Remove other properties than label and value from Combobox options in Storybook-examples, as we have limited the format for now

it-vegard committed:
Merge branch 'main' into combobox-complex-options

# Conflicts:
#	@navikt/core/react/src/form/combobox/combobox.stories.tsx
#	yarn.lock

it-vegard committed:
Array.includes does not work with objects, so should use isInList utility function. Fixes missing checkmark for selected items in FilteredOptions.

it-vegard committed:
Add changeset

it-vegard commented on .changeset/fuzzy-mangos-flow.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@navikt/ds-react": patch
```
Burde det vært en minor-oppgradering for nye features? Eller er dette for smått?

JulianNymark replied:
consensus on 'minor'?

it-vegard committed:
Remove navds-dependencies in playroom that break playwright tests. These are outdated/unsynced, and will be inherited from parent package.json instead.

it-vegard committed:
Merge branch 'main' into combobox-complex-options

JulianNymark commented on @navikt/core/react/src/form/combobox/combobox.test.tsx:
```diff
@@ -3,7 +3,7 @@ import { render, screen } from "@testing-library/react";
 import React, { useId } from "react";
 import { act } from "react-dom/test-utils";
-import { describe, expect, test } from "vitest";
+import { describe, expect, test, vi } from "vitest";
```
import {it, ... } from "vitest"; 

it somehow uses the jest one without error 🤔

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -10,21 +10,18 @@ import React, {
 import useVirtualFocus, { VirtualFocusType } from "./useVirtualFocus";
 
 type FilteredOptionsProps = {
   children: any;
```
Bør kunne bruke `React.ReactNode` som type her? (Har aversjon mot `any` 😝)
```suggestion
  children: React.ReactNode;
```

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -76,7 +73,7 @@ export const FilteredOptionsProvider = ({
   const { customOptions } = useCustomOptionsContext();
 
-  const filteredOptions = useMemo(() => {
+  const filteredOptions = useMemo<ComboboxOption[]>(() => {
```
Tror ikke det er nødvendig å definere typen eksplisitt her
```suggestion
  const filteredOptions = useMemo(() => {
```

it-vegard replied:
Hmm, det var det på et tidspunkt, men det var sikkert underveis i en refaktorering.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -43,17 +43,18 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
+            (item) =>
+              item.label.toLocaleLowerCase() === text.toLocaleLowerCase(),
           );
         };
```
Her passer det kanskje bedre med `.some()`
```suggestion
        const isTextInSelectedOptions = (text: string) =>
          selectedOptions.some(
            (option) =>
              option.label.toLocaleLowerCase() === text.toLocaleLowerCase(),
          );
```

it-vegard committed:
Do not try to remove an undefined option

it-vegard committed:
"it" is from jest. Use "test" instead.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -75,6 +76,52 @@ MultiSelect.args = {
+  return (
+    <>
+      <UNSAFE_Combobox
+        id={id}
```
Kanskje det er like greit med en hardkodet ID?

HalvorHaugan commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -75,6 +76,52 @@ MultiSelect.args = {
+                selectedOptions.filter((o) => o.value !== value),
+              )
+        }
+        size={props.size}
```
Ser ut som du egentlig bare kan spread'e props'a?

HalvorHaugan commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -75,6 +76,52 @@ MultiSelect.args = {
+    { label: "Banana", value: "banana" },
+    { label: "Orange", value: "orange" },
+    { label: "Pear", value: "pear" },
+    { label: "Grape", value: "grape" },
```
For mer realistisk case burde kanskje value være noe helt annet enn label, feks. tall.

it-vegard replied:
Bra innspill! Har endret til temakoder (brukes for søknader for å rute til riktig team++).

HalvorHaugan commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -75,6 +76,52 @@ MultiSelect.args = {
+        isMultiSelect={props.isMultiSelect}
+        onToggleSelected={(value, isSelected) =>
+          isSelected
+            ? setSelectedOptions([...selectedOptions, { label: value, value }])
```
Er det noe poeng å ha `selectedOptions`-staten som objekt når label og value uansett er likt? (Tenker da på staten i storyen og ikke eventuell intern state i Combobox)

it-vegard replied:
Nei, gir ikke mening her. I hvert fall ikke etter at jeg forbedret eksempelet. Endret til strings.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -1,6 +1,23 @@
+ * A more complex version of options for the Combobox.
+ * Used for separating the label and the value of the option.
+ *
+ * May be expanded in the future with more properties for supporting icons, grouping, etc.
```
Trenger/bør vi skrive noe om hva som _kan_ bli lagt til i fremtiden? 🤔

HalvorHaugan commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -22,7 +39,7 @@ export interface ComboboxProps
   /**
    * List of options to use for autocompletion
```
`options` brukes vel ikke bare til autocomplete?

HalvorHaugan commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -91,13 +108,13 @@ export interface ComboboxProps
-   * @param option
+   * @param option - The option value
    * @param isSelected - Whether the option has been selected or unselected
    * @param isCustomOption - Whether the option comes from user input, instead of from the list
```
Man trenger ikke ha strek mellom navn og beskrivelse
```suggestion
   * @param option The option value
   * @param isSelected Whether the option has been selected or unselected
   * @param isCustomOption Whether the option comes from user input, instead of from the list
```

HalvorHaugan commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -91,13 +108,13 @@ export interface ComboboxProps
+   * @param option - The option value
    * @param isSelected - Whether the option has been selected or unselected
    * @param isCustomOption - Whether the option comes from user input, instead of from the list
    * @returns
```
```suggestion
```

it-vegard replied:
Tok bort to andre tilfeller av @returns også

HalvorHaugan commented on aksel.nav.no/website/pages/eksempler/combobox/with-complex-options.tsx:
```diff
@@ -0,0 +1,59 @@
+
+export const args = {
+  index: 1,
+  desc: "Ved å sette en grense for maks antall valgte vil brukeren få opp en beskjed om at hen ikke kan velge flere når grensen er nådd. Resterende valg vil også bli inaktive.",
```
Glemte du å endre desc?

HalvorHaugan commented on aksel.nav.no/website/pages/eksempler/combobox/with-complex-options.tsx:
```diff
@@ -0,0 +1,59 @@
+  { label: "Helicopter", value: "helicopter" },
+  { label: "Truck", value: "truck" },
+  { label: "Van", value: "van" },
+  { label: "Scooter", value: "scooter" },
```
Er veldig pirkete nå, men synes bare det er litt rart at valgene står på engelsk mens label er på norsk 😅

it-vegard replied:
Oppdaterte til å bruke temakoder her også

it-vegard committed:
Raise change level to minor, to signal the new feature

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/react/src/form/combobox/Input/Input.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Spread props instead of adding them separately

it-vegard committed:
Update @navikt/core/react/src/form/combobox/types.ts

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
JSDoc @returns doesn't do anything without a value specified, and neither of these return a value

it-vegard committed:
Options are used for multiple features, so will just skip specifying them

it-vegard committed:
Doesn't need a guaranteed unique ID for storybook examples. Use hard-coded instead.

it-vegard committed:
selectedOptions is a list of strings anyhow, so should just use the value

it-vegard committed:
More realistic example for complex options, where the option label is larger, and the SelectedOptions label uses a short code (the value)

Also made the isInList function better, by checking both label and value when checking ComboboxOptions. Pass just the value or just the label if you only want to check either of them.

it-vegard committed:
Add tests for mapToComboboxOptionArray (renamed from mapFromStringArrayToComboboxOptionArray)

it-vegard committed:
Add test for toComboboxOption and use this function in mapToComboboxOptionArray

it-vegard committed:
Update combobox unit test for complex options to use a more realistic example

it-vegard committed:
Don't say anything about what might be added in the future. That doesn't belong in the documentation

it-vegard committed:
Oppdater beskrivelsen for complex options-eksempelet (var kopiert fra annet eksempel)

it-vegard committed:
Merge branch 'main' into combobox-complex-options

# Conflicts:
#	yarn.lock

it-vegard committed:
Fix Storybook example bug where de-selecting a value selected it an extra time

KenAJoh committed:
Merge branch 'main' into combobox-complex-options

it-vegard committed:
Fix example

it-vegard committed:
Remove listing options because it breaks with allowNewValues

it-vegard committed:
Revert earlier change that always showed selected values in FilteredOptions, even if it didn't match the search text. This broke autocomplete.

it-vegard committed:
Merge branch 'main' into combobox-complex-options

it-vegard committed:
Remove unneccessary optional check

it-vegard committed:
yarn.lock

it-vegard committed:
Don't need to use toLocaleLowerCase

it-vegard committed:
Merge branch 'main' into combobox-complex-options

KenAJoh reviewed (approved):
🚀 🚀 🚀 🚀 🚀
